src/interpreter/executors/main.rs

```diff
@@ -211,9 +211,13 @@ where
             self.tx.set_receipts_root(receipts_root);
         }
 
-        // refund remaining global gas
-        let gas_refund = self.registers[REG_GGAS] * self.tx.gas_price();
+        // used bytes are non-refundable
+        let factor = self.params.gas_price_factor as f64;
+        let gas_refund = self.tx.gas_price().saturating_mul(self.registers[REG_GGAS]) as f64;
```

Comment:
saturating is fine here since we already validated the full g_gas deduction during init